homeassistant/components/tplink/config_flow.py

```diff
@@ -1,11 +1,191 @@
+            )
+        except SmartDeviceException:
+            return self.async_abort(reason="cannot_connect")
+        return await self.async_step_discovery_confirm()
+
+    async def async_step_discovery_confirm(self, user_input=None) -> FlowResult:
+        """Confirm discovery."""
+        assert self._discovered_device is not None
+        if user_input is not None:
+            return self.async_create_entry(
+                title=f"{self._discovered_device.alias} {self._discovered_device.model}",
+                data={
+                    CONF_HOST: self._discovered_ip,
+                },
+            )
+
+        self._set_confirm_only()
+        placeholders = {
+            "name": self._discovered_device.alias,
+            "model": self._discovered_device.model,
+            "host": self._discovered_ip,
+        }
+        self.context["title_placeholders"] = placeholders
+        return self.async_show_form(
+            step_id="discovery_confirm", description_placeholders=placeholders
+        )
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            if not user_input.get(CONF_HOST):
+                return await self.async_step_pick_device()
+            try:
+                device = await self._async_try_connect(
+                    user_input[CONF_HOST], raise_on_progress=False
+                )
+            except SmartDeviceException:
+                errors["base"] = "cannot_connect"
+            else:
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=f"{device.alias} {device.model}",
+                    data={
+                        CONF_HOST: user_input[CONF_HOST],
+                    },
+                )
+
+        user_input = user_input or {}
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Optional(CONF_HOST, default=user_input.get(CONF_HOST, "")): str}
+            ),
+            errors=errors,
+        )
+
+    async def async_step_pick_device(self, user_input=None) -> FlowResult:
+        """Handle the step to pick discovered device."""
+        if user_input is not None:
+            device_id: str = user_input[CONF_DEVICE]
+            mac = device_id.split(" ")[-1]
+            device: SmartDevice = self._discovered_devices[mac]
+            await self.async_set_unique_id(mac, raise_on_progress=False)
+            self._abort_if_unique_id_configured()
+            return self.async_create_entry(
+                title=f"{device.alias} {device.model}",
+                data={
+                    CONF_HOST: device.host,
+                },
+            )
+
+        configured_devices = {
+            entry.unique_id
+            for entry in self._async_current_entries()
+            if entry.unique_id
```

Comment:
exclude `if entry.unique_id is None or entry.unique_id == DOMAIN:` once its broken out into a function